[FIX] project: allows to quick create record from My Tasks kanban

huca-odoo opened the pull request:
The refactor of the kanban quick create implemented in
https://github.com/odoo/odoo/pull/117449 introduced a traceback when
creating a task from the My Tasks kanban view due to the presence of
'default_personal_stage_type_ids' key in the context.

This commit fix that by removing this key from the context in the
customization of the kanban view for project.task model. It also add a
js tour test to ensure that personal tasks can be created from the quick
create form in the My Tasks kanban view.

Steps
=====
- Install module project
- Go to the view "Tasks -> My Tasks"
- In the main Kanban View click on "New"
- Enter a Task title and clicl on "ADD"

Issue
=====
A traceback appears indicating that the value of the field user_id is
missing for the model project.task.stage.personal and the task is not
created.

Cause
=====
In kanban and list views, when the records a grouped by a given field, a
key is added in the context when creating a record from one of the
resulting categories. This key is of the form:
`default_fieldGroupedBy: current_category_value`

In the case of the view My Tasks, the records are grouped by their
fields personal_stage_type_ids (M2M). Therefore, the key
default_personal_stage_type_ids will results in a write in that M2M
table when creating the record. But the way personal stages are managed
in project is as confusing as it is mindblowing and the M2M table
storing the relation between task and personal stages is also the table
that store records from the model project.task.stage.personal. And this
model also as a field user_id that is required. Therefore
at creation of the project.task record with that key, the orm will miss
a user_id for the newly created project.task.stage.personal.

Fix
===
The correct way to add/modify a record project.task by setting its
personal stage is actually to use the already existing methods of
project.task, i.e. _inverse_personal_stage_type_id. Therfore, instead of
modifying directly the M2M relation, update the computed M2O field and
let the inverse method to the job of personal stage assignment.

To do that, the value contained in the context for the M2M relation is
copied in the write value of the M2O (personal_stage_type_id) and the
key in the context corresponding to the M2M relation is deleted.

task-3281005

auon-odoo commented on addons/project/static/src/views/project_task_kanban/project_task_kanban_record_quick_create.js:
```diff
@@ -0,0 +1,10 @@
+
+export class ProjectTaskKanbanRecordQuickCreate extends KanbanRecordQuickCreate {
+    async getQuickCreateProps(props) {
+        await super.getQuickCreateProps(props);
+        delete this.quickCreateProps.context.default_personal_stage_type_ids
```
```suggestion
        const res = await super.getQuickCreateProps(props);
        delete this.quickCreateProps.context.default_personal_stage_type_ids;
        return res;
```
Shouldn't we return something? :thinking:

auon-odoo reviewed (changes requested):
Thank you for the fix! :fire: 
Could you please write a commit message with the following structure:
- Steps (how to reproduce the issue)
- Issue (what is happening vs what is expected)
- Cause (technical explanation)
- Fix (in consequence of the cause, why this way to fix it is best)

Of course, sometimes, some of it is obvious. But generally, it provides the reviewer with a much needed context. (And personally, writing this "report" actually made me rethink my fix a few times.)
In this particular case, it is not obvious to me why having a key in the context would result in a traceback, nor why it is not needed and so can be deleted without consequences.

Thanks again :pray:

huca-odoo commented:
> Thank you for the fix! fire Could you please write a commit message with the following structure:
> 
> * Steps (how to reproduce the issue)
> * Issue (what is happening vs what is expected)
> * Cause (technical explanation)
> * Fix (in consequence of the cause, why this way to fix it is best)
> 
> Of course, sometimes, some of it is obvious. But generally, it provides the reviewer with a much needed context. (And personally, writing this "report" actually made me rethink my fix a few times.) In this particular case, it is not obvious to me why having a key in the context would result in a traceback, nor why it is not needed and so can be deleted without consequences.
> 
> Thanks again pray

I've updated the commit message and changed the implementation as it wasn't working really well... Thanks for your comment!

auon-odoo reviewed (approved):
Lgtm! Amazing commit message :heart_eyes: :heart_eyes:

xavierbol reviewed (approved):
LGTM, thanks for your work. 🙂 

robodoo r+
